### PR TITLE
feat(sim): pool topology and disaggregation decider interface (PD PR1)

### LINKED ISSUE #794: feat(sim): PD disaggregation — pool topology and decider interface (PR1)
Part of #793 — Prefill-Decode Disaggregation Simulation (Tracking Issue)

**Depends on:** nothing (first in series)
**Blocked by:** —

## Value

Introduces all PD domain types with zero behavior change. Pure additive — no existing behavior modified. Compiles and tests independently.

## New files

| File | Contents |
|------|----------|
| `sim/disaggregation.go` | `DisaggregationDecision` struct, `DisaggregationDecider` interface, `NeverDisaggregate`, `AlwaysDisaggregate`, `NewDisaggregationDecider()` factory. Parameterized deciders (`PrefixThresholdDecider`, `DirectToDecodeDecider`) deferred to later issues. |
| `sim/disaggregation_test.go` | Tests for interface, both implementations, factory, INV-9 compliance |
| `sim/cluster/pool.go` | `PoolRole` enum (`PoolRolePrefill`, `PoolRoleDecode`), `ValidatePoolTopology()`, `BuildPoolMembership()`, `BuildPoolMembershipFromIndices()`, `FilterSnapshotsByPool()` |
| `sim/cluster/pool_test.go` | Validation, membership building, snapshot filtering |
| `sim/cluster/parent_request.go` | `ParentRequest` struct with all lifecycle timestamp fields, `NewParentRequest()` |

## Modified files

| File | Change |
|------|--------|
| `sim/bundle.go` | Add `validDisaggregationDeciders` map, `IsValidDisaggregationDecider()`, `ValidDisaggregationDeciderNames()`. Register all 4 decider names (`""`, `"never"`, `"always"`, `"prefix-threshold"`, `"direct-to-decode"`). |

## Key design decisions

- `DisaggregationDecider.Decide()` takes `*Request` and must not read `Request.OutputTokens` (INV-9 oracle boundary)
- `ParentRequest` is introduced here because it is pure data with no cluster-state dependencies
- Parameterized deciders (`PrefixThresholdDecider`, `DirectToDecodeDecider`) are registered in the valid names map but construction panics with a helpful message directing callers to the typed constructors

## Estimated size

~250 lines, 6 files

### PR DESCRIPTION
**Pull request by namasl**
## Summary

First PR in the PD disaggregation series (tracking: #793). Pure additive domain types with **zero behavior change** to existing functionality. All existing tests continue to pass byte-identically.

- **`sim/disaggregation.go`**: `DisaggregationDecision` struct, `DisaggregationDecider` interface, `NeverDisaggregate` and `AlwaysDisaggregate` implementations, `NewDisaggregationDecider` factory (panics with typed-constructor guidance for parameterized deciders)
- **`sim/disaggregation_test.go`**: Tests for interface compliance, both implementations, factory behavior, and INV-9 oracle boundary (decisions are invariant to `OutputTokens`)
- **`sim/cluster/pool.go`**: `PoolRole` enum (prefill/decode), `ValidatePoolTopology`, `BuildPoolMembership`, `BuildPoolMembershipFromIndices`, `FilterSnapshotsByPool`
- **`sim/cluster/pool_test.go`**: Tests for topology validation, membership construction, snapshot filtering, and string representation
- **`sim/cluster/parent_request.go`**: `ParentRequest` data container with lifecycle timestamp fields and `NewParentRequest` constructor
- **`sim/bundle.go`**: Registers all 4 disaggregation decider names in validation map; exports `IsValidDisaggregationDecider` and `ValidDisaggregationDeciderNames`

Closes #794

## Test plan

- [x] `go build ./...` passes
- [x] `go test ./...` passes (all packages)
- [x] New tests cover: NeverDisaggregate, AlwaysDisaggregate, factory dispatch, factory panics for unknown/parameterized names, ValidatePoolTopology edge cases, BuildPoolMembershipFromIndices, FilterSnapshotsByPool (prefill/decode/empty/order), PoolRole.String(), INV-9 oracle boundary

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: sim/disaggregation.go [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+	case "", "never":
+		return &NeverDisaggregate{}
+	case "always":
+		return &AlwaysDisaggregate{}
+	case "prefix-threshold":
+		panic("use NewPrefixThresholdDecider(threshold, blockSize) to construct prefix-threshold decider")
+	case "direct-to-decode":
+		panic("use NewDirectToDecodeDecider(threshold) to construct direct-to-decode decider")
+	default:
+		panic(fmt.Sprintf("unhandled disaggregation decider %q", name))
```
Comment (sriumcp) [inline sim/disaggregation.go]: **🟡 Improve the `default` panic message to explain the sync issue**

The `default` branch is a valid safety net — it catches the case where someone adds a name to `validDisaggregationDeciders` in `bundle.go` but forgets to add the matching `case` here. However, the current message just says `"unhandled disaggregation decider"`, which doesn't explain _why_ this happened or what to do.

```go
default:
    panic(fmt.Sprintf(
        "disaggregation decider %q is registered in validDisaggregationDeciders (bundle.go) " +
        "but has no case in NewDisaggregationDecider; add a case here",
        name))
```

This makes the eventual error self-diagnosing for the next developer who adds a new decider name.

File: sim/cluster/parent_request.go [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+	// CompletionTime has two meanings depending on outcome:
+	//   - Successful decode: set by detectDecodeCompletions when the decode sub-request
+	//     finishes its last step. CompletionTime == actual decode completion time.
+	//   - Dropped at decode KV allocation: set to the DecodeRoutingEvent time (the point
+	//     when the drop was detected). CompletionTime < actual decode time (which never
+	//     happened). Use DecodeInstanceID == "" to distinguish dropped requests.
+	CompletionTime int64
+
+	// Instance assignment
+	PrefillInstanceID string
```
[the remark this answers is not in the document]
Comment (namasl) [reply sim/cluster/parent_request.go]: Fixed in 6223464. Changed both fields to `InstanceID`.

File: sim/cluster/pool_test.go [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+				if _, ok := membership[string(id)]; ok {
+					t.Errorf("unassigned instance %q should not appear in membership", id)
+				}
+			}
+		})
+	}
+}
+
+// instanceName returns the instance ID string for index i, mirroring NewClusterSimulator convention.
+func instanceName(i int) string {
```
[the remark this answers is not in the document]
Comment (namasl) [reply sim/cluster/pool_test.go]: Fixed in 6223464. Replaced with `strconv.Itoa`, removed the custom `itoa` helper entirely.

**Review by sriumcp — APPROVED** [review]
## Review pass ✅

All 7 inline issues from the previous review are resolved in the fix commit. Tests pass, 0 lint issues.

### What was fixed

| Issue | Status |
|-------|--------|
| 🔴 Structural test (`%T` type assertion) | ✅ Replaced with `Decide()` behavioral assertion; `fmt` import removed |
| 🔴 Zero test coverage on `NewParentRequest` | ✅ `parent_request_test.go` added — ceiling-division (7 boundary cases), sub-req ID convention, arrival-time propagation, panic guards for 0 and negative block size |
| 🟠 `BuildPoolMembershipFromIndices` `total` unused | ✅ Bounds check added; phantom-ID generation is now impossible |
| 🟠 `BuildPoolMembership` cryptic panic | ✅ Domain-aware panic message added |
| 🟠 `PrefillInstanceID`/`DecodeInstanceID` type | ✅ Changed to `InstanceID` |
| 🟡 Factory `default` panic message | ✅ Now explains the validation-map/switch sync failure |
| 🟡 Custom `itoa` in test | ✅ Replaced with `strconv.Itoa` |

### One open item before PR2

**Design doc** (issue #6 from the prior review) — the module boundaries introduced here (`DisaggregationDecider` interface, `ParentRequest` lifecycle tracker, pool topology) warrant a design doc in `docs/plans/` before PR2 wires them into `ClusterSimulator`'s event pipeline. This PR itself is unblocked since it's pure-additive types, but please create the design doc before PR2 is opened.
